src/core_read_parameter_file.c

```diff
@@ -2,217 +2,260 @@
 #include <stdlib.h>
 #include <string.h>
 #include <math.h>
+#include <ctype.h> /* for isblank()*/
 
 #include "core_allvars.h"
 #include "core_mymalloc.h"
 
+enum datatypes {
+    DOUBLE = 1,
+    STRING = 2,
+    INT = 3
+};
+
+#define MAXTAGS          300  /* Max number of parameters */
+#define MAXTAGLEN         50  /* Max number of characters in the string param tags */
+
 int read_parameter_file(const int ThisTask, const char *fname, struct params *run_params)
 {
     int errorFlag = 0;
     int *used_tag = 0;
-    char my_treetype[MAX_STRING_LEN], my_outputformat[MAX_STRING_LEN];
+    char my_treetype[MAX_STRING_LEN], my_outputformat[MAX_STRING_LEN], my_forest_dist_scheme[MAX_STRING_LEN];
     /*  recipe parameters  */
     int NParam = 0;
-    char ParamTag[MAXTAGS][50];
+    char ParamTag[MAXTAGS][MAXTAGLEN + 1];
     int  ParamID[MAXTAGS];
     void *ParamAddr[MAXTAGS];
 
+    /* Ensure that all strings will be NULL terminated */
+    for(int i=0;i<MAXTAGS;i++) {
+        ParamTag[i][MAXTAGLEN] = '\0';
+    }
+    
     NParam = 0;
 
     if(ThisTask == 0) {
         printf("\nreading parameter file:\n\n");
     }
 
-    strcpy(ParamTag[NParam], "FileNameGalaxies");
+    strncpy(ParamTag[NParam], "FileNameGalaxies", MAXTAGLEN);
     ParamAddr[NParam] = run_params->FileNameGalaxies;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "OutputDir");
+    strncpy(ParamTag[NParam], "OutputDir", MAXTAGLEN);
     ParamAddr[NParam] = run_params->OutputDir;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "TreeType");
+    strncpy(ParamTag[NParam], "TreeType", MAXTAGLEN);
     ParamAddr[NParam] = my_treetype;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "TreeName");
+    strncpy(ParamTag[NParam], "TreeName", MAXTAGLEN);
     ParamAddr[NParam] = run_params->TreeName;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "SimulationDir");
+    strncpy(ParamTag[NParam], "SimulationDir", MAXTAGLEN);
     ParamAddr[NParam] = run_params->SimulationDir;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "FileWithSnapList");
+    strncpy(ParamTag[NParam], "FileWithSnapList", MAXTAGLEN);
     ParamAddr[NParam] = run_params->FileWithSnapList;
     ParamID[NParam++] = STRING;
 
-    strcpy(ParamTag[NParam], "LastSnapShotNr");
+    strncpy(ParamTag[NParam], "LastSnapShotNr", MAXTAGLEN);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  Risk is low because the source is a constant string.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)

Reply:
All these have been fixed by NULL terminating every string at the very top.